[DX-1047] Upgrade Guides

dcs3spp opened the pull request:
### **User description**
### For internal users - Please add a Jira DX PR ticket to the subject!
<br>

[DX-1047](https://tyktech.atlassian.net/browse/DX-1047)
---

### Preview Link
<!-- Add a direct preview link to make it easier for the reviewer to review. Best to add a few direct links to the pages in the PR -->
[Draft Overview](https://deploy-preview-4102--tyk-docs.netlify.app/docs/nightly/upgrading-tyk)
[Upgrade PreRequisites](https://deploy-preview-4102--tyk-docs.netlify.app/docs/nightly/developer-support/upgrading-tyk/upgrade-prerequisites)
[Upgrade Strategy](https://deploy-preview-4102--tyk-docs.netlify.app/docs/nightly/developer-support/upgrading-tyk/upgrade-strategy)
[Self Managed Go Plugins](https://deploy-preview-4102--tyk-docs.netlify.app/docs/nightly/developer-support/upgrading-tyk/self-managed/go-plugins)
[Self Managed Debian](https://deploy-preview-4102--tyk-docs.netlify.app/docs/nightly/developer-support/upgrading-tyk/self-managed/linux-distributions/self-managed-deb)
[Self Managed RPM](https://deploy-preview-4102--tyk-docs.netlify.app/docs/nightly/developer-support/upgrading-tyk/self-managed/linux-distributions/self-managed-rpm/)
[Hybrid](https://deploy-preview-4102--tyk-docs.netlify.app/docs/nightly/developer-support/upgrading-tyk/cloud/hybrid/)
[Cloud Saas](https://deploy-preview-4102--tyk-docs.netlify.app/docs/nightly/developer-support/upgrading-tyk/cloud/cloud-saas)

### Description
<!-- 1. Describe your changes in detail. Why is this change required? What problem does it solve?
     2. Please explain the change to the reviewer. Pay special attention to changes that are hard to spot, 
       for example:
       2.1. Name change in the file name or directory name - please flag it out since it’s hard 
            to compare text after such a change 
       2.2. Terminology change - flag it out to make sure the reviewer is aware before approving
     3. @mentions of the person to review the proposed changes. They need to be able to know the topic well in order to approve it.

-->
<br>
This PR adds the template structure for adding the Go plugins upgrade documents.
For quicker review and release a separate PR will be created for the other deployment upgrade guides

#### Screenshots (if appropriate)
<br>

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply -->

- [x] I have added a **preview link** to the PR description.
- [x] I have [reviewed the guidelines](https://github.com/TykTechnologies/tyk-docs/blob/master/CONTRIBUTING.md) for contributing to this repository.
- [x] I have [read the technical guidelines](https://github.com/TykTechnologies/tyk-docs/blob/master/CONTRIBUTING-TECHNICAL-GUIDE.md) for contributing to this repository.
- [x] Make sure you have started *your change* off *our latest `master`*.
- [x] I **labelled** the PR
<!-- Label your PR according to the type of changes that your code introduces. This ensures that we know how/when to publish the PR. These are the options:
- Fixing typo (please merge to production) - add the label `now`
- Documenting a new feature (please merge to production) - add the label `now`
- Documentation for future release (please do not merge to production) - add label `future release` and label of the release
- [Something else (please add if needs merging to production or not) - add label according to the above logic -->


[DX-1047]: https://tyktech.atlassian.net/browse/DX-1047?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ


___

### **PR Type**
documentation


___

### **Description**
- Added a new guide for upgrading Go plugins, including detailed steps for different upgrade paths and commands for recompiling plugins using Docker.
- Added guidelines for pre-upgrade considerations, covering upgrade strategy, backups, Go plugins, and Linux distributions.
- Introduced a new document explaining rolling and blue-green upgrade strategies with detailed steps.
- Overhauled the main upgrading Tyk document, adding sections on standards and recommendations and providing a table of contents for different upgrade guides.
- Updated the menu structure to include new upgrade guides and organized them by deployment model.


___



### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Documentation
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>go-plugins.md</strong><dd><code>Added Go Plugin Upgrade Guide with Detailed Steps</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tyk-docs/content/developer-support/upgrading-tyk/go-plugins.md
<li>Added a new guide for upgrading Go plugins.<br> <li> Included detailed steps for different upgrade paths based on Tyk <br>versions.<br> <li> Provided commands for recompiling plugins using Docker.<br>


</details>
    

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk-docs/pull/4102/files#diff-d55bcf5bf73ad329b0286448382a36eb661c4fca95e55ae7b42a0d88f6abfa3c">+105/-0</a>&nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>upgrade-guidelines.md</strong><dd><code>Added Upgrade Guidelines for Pre-Upgrade Considerations</code>&nbsp; &nbsp; </dd></summary>
<hr>

tyk-docs/content/developer-support/upgrading-tyk/preparations/upgrade-guidelines.md
<li>Added guidelines for pre-upgrade considerations.<br> <li> Included sections on upgrade strategy, backups, Go plugins, and Linux <br>distributions.<br> <li> Provided a checklist for upgrade preparations.<br>


</details>
    

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk-docs/pull/4102/files#diff-1a660f5104cc950666aa91f5bd492485a4777b75eea508874c060c183c47c38c">+54/-0</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>upgrade-strategies.md</strong><dd><code>Added Document on Rolling and Blue-Green Upgrade Strategies</code></dd></summary>
<hr>

tyk-docs/content/developer-support/upgrading-tyk/preparations/upgrade-strategies.md
<li>Added a new document explaining rolling and blue-green upgrade <br>strategies.<br> <li> Included detailed steps for both strategies.<br>


</details>
    

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk-docs/pull/4102/files#diff-7e8030c0c7755559bef672ca500928f5132594a9fe82498da9660fe32088d1d8">+32/-0</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>upgrading-tyk.md</strong><dd><code>Overhauled Main Upgrading Tyk Document with New Structure</code></dd></summary>
<hr>

tyk-docs/content/upgrading-tyk.md
<li>Overhauled the main upgrading Tyk document.<br> <li> Added sections on standards and recommendations.<br> <li> Provided a table of contents for different upgrade guides.<br>


</details>
    

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk-docs/pull/4102/files#diff-221a7f51df6259b8406c9f416126d25f2fb55b17cac68b4227a85efc90ae751b">+55/-243</a></td>
</tr>                    
</table></td></tr><tr><td><strong>Configuration changes
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>menu.yaml</strong><dd><code>Updated Menu Structure for New Upgrade Guides</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tyk-docs/data/menu.yaml
<li>Updated the menu structure to include new upgrade guides.<br> <li> Added links to new documents on upgrade guidelines and strategies.<br> <li> Organized upgrade guides by deployment model.<br>


</details>
    

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk-docs/pull/4102/files#diff-a6789c66bfa0660f7fdbe4796002ab4777677bae59a0276689c840100ff0b0b2">+79/-4</a>&nbsp; &nbsp; </td>
</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

dcs3spp commented:
@allaneyles To facilitate reviewing the PR queue we are closing some PRs that are in draft status, e.g. awaiting docker images, videos, images etc. You can still update the PR and then open it again when ready for review.

cc: @letzya

dcs3spp reviewed (commented):
Hi @Monica-Cedeno I have had a read through self-managed-rpm section and content looks good! Review comments so far:

Is there some steps missing in bundling the plugins section:
- Upload the plugin bundle to a webserver
- [Configure Gateway](https://tyk.io/docs/plugins/how-to-serve-plugins/plugin-bundles/#global-parameters) to download plugins from the webserver
- Add the plugin bundle filename (e.g. my-plugin-bundle.zip)  as the ID in the Advanced options for an API in the Dashboard screen

Is the content the same for the _custom-plugins upgrade guide_ and the _upgrade strategy_ guide in self-managed-deb and self-managed-rpm?

For command terminal / shell code formatting examples use _bash_, e.g:

```suggestion
```bash
go get github.com/TykTechnologies/tyk@ffa83a27d3bf793aa27e5f6e4c7106106286699d
# In Gateway version 5.1, the Gateway and plugins transitioned to using Go modules builds and don’t use Go mod vendor anymore
go mod tidy
```
```

For JSON examples use _json_, e.g.:

```suggestion
```json
{
    "key": "value"
}
```
```

Fadzli-fast commented:
Hey @dcs3spp, allow to explain some parts from the guide

**Self-Managed RPM & DEB Guides
What is the action of the target package version step in the guide?** 

- This section is for users to identify the version that they are upgrading into after their initial discussion on which versions to upgrade to. Eg, the customer has decided to upgrade to version 5.2.5, isit available for their specific OS Rhel(el/7, el/8) or Ubuntu focal, bionic etc. 

**Images of building bundles, available versions of Tyk components are difficult to read**

- Will it be better removed? as its quite straight forward? This images are usually requested by APAC customers. eg Do you have an example of how it looks like?

**Upgrade Strategy
What is the deployment document that you wish to link to?**

- I believe its supposed to point to the [deployment page](https://tyk.io/docs/getting-started/installation/)

**Self-Managed Custom Go Plugins
Should the bundling guide be included for each go plugin initialisation sections rather than just the section for versions > 5.1?**

This is on assumption **IF** you are using bundles to ship your plugins, to avoid confusion I feel its better if the [bundling section ](https://tyk.io/docs/plugins/how-to-serve-plugins/plugin-bundles/#global-parameters) is added into the guide rather then creating a bunch of different instructions? eg **If you are using bundles to ship your plugins refer to link**

dcs3spp reviewed (commented):
Many thanks @Fadzli-fast for answering the review questions!!!

I have created this [child PR](https://github.com/TykTechnologies/tyk-docs/pull/4195) based with new structure for merging into the original PR. I also removed the plugin bundles section from self-managed guides and replaced with link to existing documentation. @Fadzli-fast @Monica-Cedeno @nerdydread could you please **review** and if ok **approve** it so that I can merge it into the original DX-1047 PR?

The remainder of DX review summary is:

**Self-Managed Custom Go Plugins**
- Section headings do not match up with the version range in the table

**Pre-upgrade checks**
- For this line "Is the targeted version available on packagecloud.io or the intended version" Should it read "is the targeted version available on packagecloud.io or in local package repository?" 

**Self Managed RPM**
- A new image is required for available versions of Tyk components. Existing [image](https://deploy-preview-4102--tyk-docs.netlify.app/docs/nightly/developer-support/upgrading-tyk/go-plugins/self-managed-rpm/#check-upgradable-packages) is difficult to read.

**Self Managed Hybrid** 
- Thanks for uploading, the self managed hybrid guide is now available for me to review.

**Cloud Saas**
- Add supporting images to the guide
- Does the Go Plugins table match the section headings?
- Add link to guide for how to upgrade the Tyk Data Plane (Gateway)

nerdydread commented:
@dcs3spp Thank you for the feedback, changes have been applied in the following [commit](https://github.com/TykTechnologies/tyk-docs/commit/5a513cc8f27fe5dfbfeba92ec72d3adceef971b8).
> **Cloud Saas**
> 
>     * Add supporting images to the guide
> 
>     * Does the Go Plugins table match the section headings?
> 
>     * Is there a guide that can link to for how to upgrade the Tyk Data Plane (Gateway)
> 

- Added images for showing an example of a bundle’s contents.
- The table and section headings have been updated to match.  
- Removed the steps and linked to the existing guide for upgrading the Tyk Data Plane.  Added links to other existing docs for certain steps as some things were already well documented and this would cut down on the amount of duplication.

dcs3spp commented:
@allaneyles @Fadzli-fast @Monica-Cedeno @nerdydread I have reviewed the Hybrid upgrade guide and also updated the Hybrid and Cloud SaaS guides for formatting, added links and sub-section headings. In the Hybrid Gateway document I have also updated the definition of hybrid Gateway so that it relates to MongoDB or PostgreSQL.

The following is a summary of the DX review and a peer review should also be performed before this is released.

[**Self-Managed Custom Go Plugins**](http://localhost:1313/docs/nightly/developer-support/upgrading-tyk/self-managed/go-plugins/)
- Section headings do not match up with the version range in the table

[**Self Managed RPM**](https://deploy-preview-4102--tyk-docs.netlify.app/docs/nightly/developer-support/upgrading-tyk/self-managed/linux-distributions/self-managed-rpm/)
- A new image is required for available versions of Tyk components. Existing [image](https://deploy-preview-4102--tyk-docs.netlify.app/docs/nightly/developer-support/upgrading-tyk/go-plugins/self-managed-rpm/#check-upgradable-packages) is difficult to read.

[**Hybrid** - Moved into cloud section](https://deploy-preview-4102--tyk-docs.netlify.app/docs/nightly/developer-support/upgrading-tyk/cloud/hybrid/)
- Is definition of hybrid correct? Is hybrid also where client hosts a Gateway in 3rd party cloud provider or on their own server?
- Should the plugins be updated on the Management Gateway also?
- Go plugins section has image of plugin bundle files with year of 1979
- Go plugins section step 5 image needed for path 1 & path 3 to match version of upgrade represented by section heading
- Path 1 & Path 3 manifest.json file example does not contain both plugin .so files but path 2 does, why? Should this be updated for paths 1 & 3 to include both plugin .so files?
- Path 3 of Go Plugins: Are steps 7 & 10 duplicated?
- Go Plugins: Steps 7, 8, 9 & 10 are in different order across upgrade paths. Update with correct consistent recommended ordering.
- In Go plugins path 1 - 3 should the link for upgrading Hybrid Gateway(s) refer to this [section](https://deploy-preview-4102--tyk-docs.netlify.app/docs/nightly/developer-support/upgrading-tyk/cloud/hybrid/#3-upgrade-your-data-plane-hybrid-gateways)? I have added this as part of refactoring.
- Update Go plugin path section headings to match table, similar to Cloud SaaS.

[**Cloud Saas**](https://deploy-preview-4102--tyk-docs.netlify.app/docs/nightly/developer-support/upgrading-tyk/cloud/cloud-saas/)
- Go plugins section has image of plugin bundle files with year of 1979
- Go plugins section step 5 image needed for path 1 & path 3 to match version of upgrade represented by section heading
- Path 1 & Path 3 manifest.json file example does not contain both plugin .so files. Should this be updated for paths 1 & 3 to include both plugin .so files?
- Steps 7, 8 & 9 are in different order of Go plugins for [Paths 1 - 3](https://deploy-preview-4102--tyk-docs.netlify.app/docs/nightly/developer-support/upgrading-tyk/cloud/cloud-saas/#path-1) Should they be the same order?
- In go plugins paths 1-3 add a link for the step that involves upgrading to the Cloud Data Plane. Should it link to this [section](https://deploy-preview-4102--tyk-docs.netlify.app/docs/nightly/developer-support/upgrading-tyk/cloud/cloud-saas/#upgrading-cloud-data-planes)? I have added the link and double checking for code review
- Path 3 (Upgrade From Current Version >= 4.1.0 To Target Version >= 5.1.0) gives an example for upgrading to _Target Version > v4.20 and < v5.1.0_. Should this example be removed since it does not match the scenario represented by the path 3 section heading?

Are the Go plugins steps the same for Hybrid and Cloud Saas, or is it safer to have them as separate?

Fadzli-fast commented:
Hey @dcs3spp,

I have updated the ones below

```
[Self-Managed Custom Go Plugins](http://localhost:1313/docs/nightly/developer-support/upgrading-tyk/self-managed/go-plugins/)

Section headings do not match up with the version range in the table
```

Already changed it to be consistent with the other guides to avoid confusion.

```
[Self Managed RPM](https://deploy-preview-4102--tyk-docs.netlify.app/docs/nightly/developer-support/upgrading-tyk/self-managed/linux-distributions/self-managed-rpm/)

A new image is required for available versions of Tyk components. Existing [image](https://deploy-preview-4102--tyk-docs.netlify.app/docs/nightly/developer-support/upgrading-tyk/go-plugins/self-managed-rpm/#check-upgradable-packages) is difficult to read.
```

I have changed the screenshots to be more easier to read focusing on the example printout of the command.

Also made some change under databases with a [note](https://deploy-preview-4102--tyk-docs.netlify.app/docs/nightly/developer-support/upgrading-tyk/self-managed/linux-distributions/self-managed-rpm/#databases) stating that we dont own them and this a guide only.

Also removed screenshots for mongo restore as it wasnt helpful plus its hard to see.

dcs3spp commented:
Thanks @Fadzli-fast Images look much better in the linux Debian & Redhat section. I like the idea of adding the  note in databases backup!!!

dcs3spp commented:
@allaneyles @Fadzli-fast @Monica-Cedeno @nerdydread [Feedback](https://github.com/TykTechnologies/tyk-docs/pull/4102#issuecomment-1991875347) remaining is Hybrid & Cloud Saas

nerdydread commented:
Hey @dcs3spp I've made the following updates based on your feedback:
  
[Hybrid - Moved into cloud sub-menu](https://deploy-preview-4102--tyk-docs.netlify.app/docs/nightly/developer-support/upgrading-tyk/cloud/hybrid/)
- Is definition of hybrid correct? Is hybrid also where client hosts a Gateway in 3rd party cloud provider or on their own server?
I was copying/pasting between the SaaS and Hybrid document as they’re similar and missed some of the updates in my first pass.  This has been updated accordingly.
- Should the plugins be updated on the Management Gateway also?
No, the plugins aren’t needed on the Management Gateway as this Gateway doesn’t proxy any API traffic.
- Go plugins section has image of plugin bundle files with year of 1979
This is a result of the zip utility I used, it stripped the data modified value.  I’ve gone back and modified the Date Modified for each of the images so as to not be a source of confusion.  
- Go plugins section step 5 image needed for path 1 & path 3 to match version of upgrade represented by section heading
Updated this accordingly.
- Path 1 & Path 3 manifest.json file example does not contain both plugin .so files but path 2 does, why? Should this be updated for paths 1 & 3 to include both plugin .so files?
We introduced our [Go Plugin Loader](https://tyk.io/docs/product-stack/tyk-gateway/release-notes/version-4.1/#go-plugin-loader) functionality in 4.1.0 that allowed users to include multiple versions of a given plugin in the same bundle and the Gateway would then select the appropriate version to load.  This allows for users to include their target version and current version plugin in the same bundle package prior to upgrading so that post-upgrade, the Gateway can pull the appropriate version automatically.  For Path 1 (< 4.1.0), you have to upgrade the Gateway then load the target version plugin.   For Paths 2 and 3, the target version plugin can be bundled and pulled down by the Gateway prior to upgrading.  Path 3 has been updated to  include both .so files.
- Path 3 of Go Plugins: Are steps 7 & 10 duplicated?
This has been updated to be more clear.
- Go Plugins: Steps 7, 8, 9 & 10 are in diff. erent order across upgrade paths. Update with correct consistent recommended ordering.
Updated accordingly, the Go Plugin Loader functionality introduced in v4.1.0 is why the steps are different between Path 1 and Paths 2-3.
- In Go plugins path 1 - 3 should the link for upgrading Hybrid Gateway(s) refer to this[ section](https://deploy-preview-4102--tyk-docs.netlify.app/docs/nightly/developer-support/upgrading-tyk/cloud/hybrid/#3-upgrade-your-data-plane-hybrid-gateways)? I have added this as part of refactoring.
Yes this should link to the existing documentation on upgrading Cloud Control Planes.  This has been updated.
- Update Go plugin path section headings to match table, similar to Cloud SaaS.
This has been updated.

[Cloud Saas](https://deploy-preview-4102--tyk-docs.netlify.app/docs/nightly/developer-support/upgrading-tyk/cloud/cloud-saas/)
- Go plugins section has image of plugin bundle files with year of 1979
This has been updated as the zip utility I used at the time stripped the Date Modified attribute.
- Go plugins section step 5 image needed for path 1 & path 3 to match version of upgrade represented by section heading
This has been updated accordingly for each path.
- Path 1 & Path 3 manifest.json file example does not contain both plugin .so files. Should this be updated for paths 1 & 3 to include both plugin .so files?
Same answer as the Hybrid section above.   
- Steps 7, 8 & 9 are in different order of Go plugins for[ Paths 1 - 3](https://deploy-preview-4102--tyk-docs.netlify.app/docs/nightly/developer-support/upgrading-tyk/cloud/cloud-saas/#path-1) Should they be the same order?
Updated accordingly, the Go Plugin Loader functionality introduced in v4.1.0 is why the steps are different between Path 1 and Paths 2/3.
- In go plugins paths 1-3 add a link for the step that involves upgrading to the Cloud Data Plane. Should it link to this[ section](https://deploy-preview-4102--tyk-docs.netlify.app/docs/nightly/developer-support/upgrading-tyk/cloud/cloud-saas/#upgrading-cloud-data-planes)? I have added the link and double checking for code review
Yes this should link to the existing documentation on upgrading Cloud Control Planes.  This has been updated.
- Path 3 (Upgrade From Current Version >= 4.1.0 To Target Version >= 5.1.0) gives an example for upgrading to Target Version > v4.20 and < v5.1.0. Should this example be removed since it does not match the scenario represented by the path 3 section heading?
This has been removed.
- Are the Go plugins steps the same for Hybrid and Cloud Saas, or is it safer to have them as separate?
They’re nearly identical, the only difference really is that SaaS deployments will use an S3 Bucket for plugin bundles while Hybrid deployments will use a file server.

letzya commented:
about cloud saas >=5.1.0 - we should check if this works with 5.3 and get confirmation for it @andyo-tyk

letzya commented on tyk-docs/content/upgrading-tyk.md:
```diff
@@ -34,21 +35,24 @@ This section applies to all self-managed components, including licensed and open
+The sections below explain how to upgrade Tyk Gateway on:
+- [Docker](#docker)
+- [Kubernetes](#k8s)
+- [Linux](#linux)
```
can we put the above in the same order of the content below?

letzya commented on tyk-docs/content/upgrading-tyk.md:
```diff
@@ -181,79 +186,6 @@ Instructions for upgrading Tyk gateway. You should follow the same flow for Tyk
    Check the [helm upgrade docs](https://helm.sh/docs/helm/helm_upgrade/) for more details on the `upgrade` command.
 
-
-### Other installation choices
```
Where is this content? what is the purpose of this page? if it's an overview, shouldn't we have links to the content that was moved out of this page to another place?

Why did we leave k8s, help and docker but took out the rest?

letzya reviewed (changes requested):
per my comment below the current upgrading tyk guide needs to be looked at. Currently it feels like it's disconnected from the new content.

letzya commented on tyk-docs/content/developer-support/upgrading-tyk/upgrade-strategy.md:
```diff
@@ -0,0 +1,21 @@
+---
+title: "Upgrading Strategy"
+date: 2024-03-1
+tags: ["Upgrade Strategy" ]
```
```suggestion
tags: ["Upgrade Strategy", "Upgrading Tyk options" ]
```

letzya commented on tyk-docs/content/developer-support/upgrading-tyk/upgrade-prerequisites.md:
```diff
@@ -0,0 +1,36 @@
+
+## Next Steps {#next-steps}
+
+Consult our [upgrade strategy]({{< ref "developer-support/upgrading-tyk/upgrade-strategy" >}}) guidelines to help you decide between a [Rolling]({{< ref "developer-support/upgrading-tyk/upgrade-strategy#rolling-upgrade" >}}) or [Blue-Green]({{< ref "developer-support/upgrading-tyk/upgrade-strategy#blue-green" >}}) upgrade.
```
```suggestion
Review our [upgrade strategy guidelines]({{< ref "developer-support/upgrading-tyk/upgrade-strategy" >}}) to decide between a [Rolling]({{< ref "developer-support/upgrading-tyk/upgrade-strategy#rolling-upgrade" >}}) or [Blue-Green]({{< ref "developer-support/upgrading-tyk/upgrade-strategy#blue-green" >}}) upgrade.
```

letzya commented on tyk-docs/content/developer-support/upgrading-tyk/upgrade-prerequisites.md:
```diff
@@ -0,0 +1,36 @@
+
+| Platform         | Guide             | Description |
+| ---------------- | ---------------- | ----------- |
+| Tyk Cloud        | [Cloud SaaS]({{< ref "developer-support/upgrading-tyk/cloud/cloud-saas" >}}) | Guide for Tyk Cloud SaaS (Software As A Service) |
```
@dcs3spp Not sure I don't get this. Why we have only linux here? where's docker and k8s?

letzya commented on tyk-docs/content/developer-support/upgrading-tyk/cloud/hybrid.md:
```diff
@@ -0,0 +1,241 @@
+- Redis instance 
+- Tyk Pump (optional)
+
+After reviewing your [upgrade pre-requisites]({{< ref "developer-support/upgrading-tyk/upgrade-prerequisites" >}}), 
```
why pre-requisites and not prerequisites?

letzya commented on tyk-docs/content/developer-support/upgrading-tyk/upgrade-prerequisites.md:
```diff
@@ -0,0 +1,36 @@
+---
+title: "Upgrade Pre-Requisites"
```
it's is read more as "Pre Upgrade Considerations" than "Upgrade Pre-Requisites" to me. What do you think

letzya commented on tyk-docs/content/developer-support/upgrading-tyk/cloud/hybrid.md:
```diff
@@ -0,0 +1,241 @@
+
+
+---
+## 1. Upgrade your Control Plane
```
@dcs3spp why do we repeat it hee if it was written in the saas

letzya commented on tyk-docs/content/developer-support/upgrading-tyk/cloud/hybrid.md:
```diff
@@ -0,0 +1,241 @@
+
+ | Upgrade Path | Current Version | Target Version |
+ | ---- | --------------- | -------------- |
+ | [1](#path-1)    | < 4.1.0         | < 4.1.0        |
```
```suggestion
 | [Path 1](#path-1)    | < 4.1.0         | < 4.1.0        |
```

letzya commented on tyk-docs/content/developer-support/upgrading-tyk/cloud/hybrid.md:
```diff
@@ -0,0 +1,241 @@
+ | ---- | --------------- | -------------- |
+ | [1](#path-1)    | < 4.1.0         | < 4.1.0        |
+ | [2](#path-2)    | < 4.1.0         | \>= 4.1.0      |
+ | [3](#path-3)    | \>= 4.1.0       | \>=5.1.0       |
```
```suggestion
 | [Path 3](#path-3)    | \>= 4.1.0       | \>=5.1.0       |
```

letzya commented on tyk-docs/content/developer-support/upgrading-tyk/cloud/cloud-saas.md:
````diff
@@ -0,0 +1,212 @@
+ go mod vendor
+ ```
+
+3. Download the plugin compiler for the target version you’re upgrading to (e.g. 4.0.9).  See the Tyk Docker Hub [repo](https://hub.docker.com/r/tykio/tyk-plugin-compiler) for available versions. 
````
```suggestion
3. Download the plugin compiler for the target version you’re upgrading to (e.g. 4.0.9).  See the Tyk Docker Hub [repo](https://hub.docker.com/r/tykio/tyk-plugin-compiler/tags) for available versions. 
```

letzya replied:
```suggestion
3. Download the plugin compiler for the target version you’re upgrading to (e.g. for v5.3 `docker pull tykio/tyk-plugin-compiler:v5.3`).  See the Tyk Docker Hub [repo](https://hub.docker.com/r/tykio/tyk-plugin-compiler) for available versions. 
```

letzya commented on tyk-docs/content/developer-support/upgrading-tyk/cloud/cloud-saas.md:
````diff
@@ -0,0 +1,212 @@
+ ```
+
+3. Download the plugin compiler for the target version you’re upgrading to (e.g. 4.0.9).  See the Tyk Docker Hub [repo](https://hub.docker.com/r/tykio/tyk-plugin-compiler) for available versions. 
+4. [Compile]({{< ref "plugins/supported-languages/golang#building-the-plugin">}}) your plugin using this compiler 
````
@dcs3spp can we give a better link? more direct to the compile command?

letzya commented on tyk-docs/content/developer-support/upgrading-tyk/cloud/cloud-saas.md:
```diff
@@ -0,0 +1,212 @@
+
+## Upgrade Guide Video
+
+Please refer to our supporting [video](https://tyk-1.wistia.com/medias/t0oamm63ae) below for further guidance.
```
is it "further guidance" ?

letzya replied:
```suggestion
Please refer to our [upgrade guide video](https://tyk-1.wistia.com/medias/t0oamm63ae) below for visual guidance:
```

letzya commented on tyk-docs/content/developer-support/upgrading-tyk/cloud/cloud-saas.md:
```diff
@@ -0,0 +1,212 @@
+
+6. [Upload this bundle]({{< ref "tyk-cloud/configuration-options/using-plugins/uploading-bundle" >}}) to your configured S3 bucket.  
+7. Update the [custom_middleware_bundle]({{< ref "plugins/how-to-serve-plugins/plugin-bundles#per-api--local-parameters" >}}) field in the API Definitions of all APIs that use your plugin. The field should be updated to use the new bundle file you created in step 5.
+8. Validate that your plugin is working per your expectations as at this stage, your Gateway will be running the plugin for your current version still.  
```
```suggestion
8. Ensure that your plugin is functioning as expected, as at this stage, your Gateway will still be running the plugin for the current version.
```

letzya commented on tyk-docs/content/developer-support/upgrading-tyk/cloud/hybrid.md:
```diff
@@ -0,0 +1,241 @@
+- MongoDB 
+- Redis (Master Instance)
+- Management Gateway
+- MDCB
```
Why don't we call it "control plane"?

letzya commented on tyk-docs/content/developer-support/upgrading-tyk/cloud/hybrid.md:
```diff
@@ -0,0 +1,241 @@
+## 1. Upgrade your Control Plane
+See Tyk Guide for how to [Upgrade Control Planes]({{< ref "tyk-cloud/environments-&-deployments/managing-control-planes#upgrade-control-planes" >}})
+
+## 2. Upgrade your Go Plugins
```
@dcs3spp we need to underatand from Travis why this info is repeated in saas and hybrid. Do we have to have duplicated content?

letzya commented on tyk-docs/content/developer-support/upgrading-tyk/cloud/hybrid.md:
```diff
@@ -0,0 +1,241 @@
+
+## 3. Upgrade your Data Plane Hybrid Gateway(s){#upgrading-data-plane-hybrid-gateways}
+Follow the instructions for component deployment type:
+- **Docker**
```
now we have this section and also this page https://deploy-preview-4102--tyk-docs.netlify.app/docs/nightly/developer-support/upgrading-tyk/other-upgrade-options/docker/ - they essentially do the same thing. Can you please improve this to avoid dup of content?

letzya commented on tyk-docs/content/developer-support/upgrading-tyk/self-managed/linux-distributions/self-managed-deb.md:
```diff
@@ -0,0 +1,222 @@
+
+In a production environment, where we recommend installing the Dashboard, Gateway, and Pump on separate machines, you should upgrade components in the following sequence:
+
+1. Tyk Gateway
```
@dcs3spp The order here is not consistent with the order we currently have in our docs.

letzya reviewed (changes requested):
@dcs3spp Added a few comments. Also the order of the upgrade in self managed is not consistent. 
Could you kindly check why and re-review the whole content in holistic way to make sure the rest is consistent, the guide is flowing and the navigation makes sense (currently it needs work)

Monica-Cedeno commented:
this looks great. Thanks @letzya and @dcs3spp

letzya commented on tyk-docs/data/menu.yaml:
```diff
@@ -4208,10 +4208,81 @@ menu:
+            category: Directory
+            show: True
+            menu:
+            - title: "Upgrade Tyk On Debian"
```
```suggestion
            - title: "Debian"
```

letzya commented on tyk-docs/data/menu.yaml:
```diff
@@ -4208,10 +4208,81 @@ menu:
+              path: /developer-support/upgrading-tyk/self-managed/linux-distributions/self-managed-deb
+              category: Page
+              show: True
+            - title: "Upgrade Tyk On Redhat (Cent OS)"
```
```suggestion
            - title: "Redhat (Cent OS)"
```

letzya commented on tyk-docs/data/menu.yaml:
```diff
@@ -4208,10 +4208,81 @@ menu:
+              path: /developer-support/upgrading-tyk/self-managed/linux-distributions/self-managed-rpm
+              category: Page
+              show: True
+          - title: "Docker"
```
```suggestion
          - title: "Docker - Tyk Gateway"
```

letzya commented on tyk-docs/data/menu.yaml:
```diff
@@ -4208,10 +4208,81 @@ menu:
+            path: /developer-support/upgrading-tyk/self-managed/docker
+            category: Page
+            show: True
+          - title: "Helm Charts"
```
```suggestion
          - title: "Helm Chart - Tyk Gateway"
```

letzya commented on tyk-docs/content/developer-support/upgrading-tyk/go-plugins.md:
```diff
@@ -0,0 +1,105 @@
+---
+title: "Go Plugin Upgrade Guide"
+date: 2024-03-1
```
```suggestion
date: 2024-06-17
```

letzya commented on tyk-docs/content/developer-support/upgrading-tyk/go-plugins.md:
````diff
@@ -0,0 +1,105 @@
+Recompile your plugin with this version
+
+```bash
+docker run --rm -v `pwd`:/plugin-source \
````
```suggestion
docker run --rm -v "$(pwd)":/plugin-source \
```

letzya commented on tyk-docs/content/upgrading-tyk.md:
```diff
@@ -1,270 +1,82 @@
+This section provides guides and guidance for upgrading your Tyk installation.
+When upgrading Tyk, you need to consider the model you are on (SaaS, Self Managed, Hybrid or OSS). Then depending on
+that understand which components you should upgrade (e.g. Gateway, Pump, Dashboard, Go Plugins), while taking into
+account the deployment style you've implemented (Docker, Helm, K8S, varios Linux distributions).
```
```suggestion
account the deployment style you've implemented (Docker, Helm, K8S, various Linux distributions).
```

letzya commented on tyk-docs/content/upgrading-tyk.md:
```diff
@@ -1,270 +1,82 @@
+specifically stated in the release (and it rarely happens).
+- **Long Term Support:** Check our [versioning and long-term-support policies]({{< ref "developer-support/special-releases-and-features/long-term-support-releases" >}})
+for more details on the way we release major and minor features, patches and the support dates for each release.
+- **Preparations:** Before starting the upgrade make sure you have consulted the [preparation guidelines]({{< ref "developer-support/upgrading-tyk/preparations/upgrade-guidelines" >}})
```
```suggestion
- **Preparations:** Make sure you have consulted the [preparation guidelines]({{< ref "developer-support/upgrading-tyk/preparations/upgrade-guidelines" >}}) before starting the upgrade.
 
```

letzya commented on tyk-docs/data/menu.yaml:
```diff
@@ -4205,13 +4205,88 @@ menu:
       show: True
+      menu:
+      - title: "Overview"
+        path: /developer-support/upgrading-tyk
```
```suggestion
        path: /developer-support/upgrading-tyk/
```

letzya replied:
```suggestion
        path: /upgrading-tyk
```

letzya commented on tyk-docs/data/menu.yaml:
```diff
@@ -4205,13 +4205,88 @@ menu:
       show: True
+      menu:
+      - title: "Overview"
+        path: /developer-support/upgrading-tyk/
```
```suggestion
        path: /developer-support/upgrading-tyk
```

letzya commented:
/release to release-5.3
